src/master/heartbeat_service.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * heartbeat_service.cpp - heartbeat communication service
+ */
+
+#include "heartbeat_service.hpp"
+
+#include "heartbeat_cluster.hpp"
+#include "udp_rpc.hpp"
+
+namespace cubhb
+{
+
+  heartbeat_service::heartbeat_service (udp_server &server, cluster &cluster_)
+    : m_server (server)
+    , m_cluster (cluster_)
+  {
+    server_request_handler handler = std::bind (&heartbeat_service::handle_heartbeat, std::ref (*this),
+				     std::placeholders::_1);
+    m_server.register_handler (HEARTBEAT, handler);
+  }
+
+  void
+  heartbeat_service::handle_heartbeat (server_request &request)
+  {
+    heartbeat_arg request_arg;
+    request.get_body (request_arg);
+
+    m_cluster.handle_heartbeat (request_arg, request.get_remote_ip_address ());
+
+    // must send heartbeat response in order to avoid split-brain when heartbeat configuration changed
+    if (request_arg.is_request () && !m_cluster.hide_to_demote)
+      {
+	heartbeat_arg response_arg (false, request_arg.get_orig_hostname (), m_cluster);
+	request.get_response ().set_body (response_arg);
+      }
+  }
+
+  void
+  heartbeat_service::send_heartbeat (const cubbase::hostname_type &node_hostname)
+  {
+    heartbeat_arg request_arg (true, node_hostname, m_cluster);
```

Comment:
hb_set_net_header used to return error if cluster was not healthy (its myself is null).

what will happen if cluster is not healthy?

Reply:
myself link is set during initialization or reload of the cluster, if myself is NULL after these 2 operation the heartbeat process will fail. In this case I think it is ok to just skip send hearbeat in case myself link is NULL